rust/azure_iot_operations_services/src/leased_lock/lease.rs

```diff
@@ -0,0 +1,326 @@
+            self.current_fencing_token
+                .lock()
+                .expect("Could not lock mutex")
+                .clone_from(&state_store_response.version);
+
+            Ok(state_store_response.version.expect(
+                "Got None for fencing token. A lease without a fencing token is of no use.",
+            ))
+        } else {
+            *self
+                .current_fencing_token
+                .lock()
+                .expect("Could not lock mutex") = None;
+
+            Err(Error(ErrorKind::KeyAlreadyLeased))
+        }
+    }
+
+    /// Attempts to acquire a lease, returning if it cannot be acquired after one attempt.
+    ///
+    /// `lease_expiration` is how long the lease will remain held in the State Store after acquired, if not released before then.
+    /// `request_timeout` is the maximum time the function will wait for receiving a response from the State Store service, it is rounded up to the nearest second.
+    ///
+    /// Returns Ok with a fencing token (`HybridLogicalClock`) if completed successfully, or `Error(KeyAlreadyLeased)` if lease is not acquired.
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidArgument`](ErrorKind::InvalidArgument) if the `request_timeout` is zero or > `u32::max`
+    ///
+    /// [`struct@Error`] of kind [`ServiceError`](ErrorKind::ServiceError) if the State Store returns an Error response
+    ///
+    /// [`struct@Error`] of kind [`UnexpectedPayload`](ErrorKind::UnexpectedPayload) if the State Store returns a response that isn't valid for a `Set` request
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError) if there are any underlying errors from the command invoker
+    ///
+    /// [`struct@Error`] of kind [`KeyAlreadyLeased`](ErrorKind::KeyAlreadyLeased) if the `lease` is already in use by another holder
+    /// # Panics
+    /// Possible panic if, for some error, the fencing token (a.k.a. `version`) of the acquired lease is None.
+    pub async fn acquire(
+        &mut self,
+        lease_expiration: Duration,
+        request_timeout: Duration,
+        renewal_period: Option<Duration>,
+    ) -> Result<HybridLogicalClock, Error> {
+        // Stop auto-renewal.
+        if !self.auto_renewal_cancellation_token.is_cancelled() {
+            self.auto_renewal_cancellation_token.cancel();
+        }
+
+        let acquire_result = self
+            .internal_acquire(lease_expiration, request_timeout)
+            .await;
+
+        if renewal_period.is_some() {
+            self.auto_renewal_cancellation_token = CancellationToken::new();
+            let mut self_clone = self.clone();
+
+            tokio::task::spawn({
+                async move {
+                    loop {
+                        select! {
+                            () = self_clone.auto_renewal_cancellation_token.cancelled() => {
+                                break; // Auto-renewal is cancelled.
```

Comment:
if auto-renewal is cancelled, does the lease need to be released? Or is it desired for the lease to be released at the next lease_expiration interval?

Reply:
The auto-renewal task is cancelled in two scenarios:
1. release() is called by the user app
2. the lease client is dropped.

On the first scenario it matches what you suggest, but on the second the lease remains until it expires. My logic was that if the customer has not called to intentionally release the lock, it should remain. Maybe the app crashed and they want to get a chance to reacquire it. Does that sound like a fair use case for you?

Reply:
yep, I think this makes sense!

For the scenario where the app crashes, there won't be any errors with them calling acquire(same lease_duration) again, right? It would just behave as an early lease renewal (which you're right, this is the ideal behavior!)